gensim/similarity_learning/sl_vocab.py

```diff
@@ -0,0 +1,557 @@
+            The size of the vectors in the above Glove vectors
+
+        maxlen: int
+            The maximum possible length of a query or a document
+            This is used for padding.
+        """
+
+        if file_path is not None:
+            with open(file_path) as f:
+                self.df = pd.read_csv(f, sep='\t')
+        else:
+            raise NotImplementedError()
+
+        self.queries = list(self.df['Question'])
+        self.documents = list(self.df['Sentence'])
+        self.relations = list(self.df['Label'])
+        self.relations = [int(r) for r in self.relations]
+
+        self.embedding_dim = embedding_dim
+        self.maxlen = maxlen
+        self.word_embedding_path = word_embedding_path
+        self.word2index, self.index2word = {}, {}
+        self.word_counter = Counter()
+        self.corpus = self.queries + self.documents
+
+        self.build_vocab()
+        self.pair_list = self.get_pair_list()
+
+    def build_vocab(self):
+        """Indexes all the words and makes an embedding_matrix which
+        can be fed directly into an Embedding layer"""
+
+        logger.info("Starting Vocab Build")
+        for sentence in self.corpus:
+            sentence = self.preprocess(sentence)
+            self.word_counter.update(sentence.split())
+        for i, word in enumerate(self.word_counter.keys()):
+            self.word2index[word] = i
+            self.index2word[i] = word
+        self.vocab_size = len(self.word2index)
+        logger.info("Vocab Build Complete")
+
+        # Since the embedding layer will a different order, the indexes
+        # of the words might change. So, we'll make a revised word2index copy
+
+        revised_word2index = {}
+        revised_index2word = {}
+
+        logger.info("Building embedding index using pretrained word embeddings")
+        embeddings_index = {}
+        with open(self.word_embedding_path) as f:
+            for line in f:
+                values = line.split()
+                word = values[0]
+                coefs = np.asarray(values[1:], dtype='float32')
+                embeddings_index[word] = coefs
+
+        self.embedding_matrix = np.zeros(
+            (len(self.word2index) + 1, self.embedding_dim))
+        for word, i in self.word2index.items():
+            embedding_vector = embeddings_index.get(word)
+            if embedding_vector is not None:
+                # words not found in embedding index will be all-zeros.
+                self.embedding_matrix[i] = embedding_vector
+                revised_word2index[word] = i
+                revised_index2word[i] = word
+
+        # TODO check below, throwing slight error
+        # revise the value of index2word and word2index
+        # self.word2index = revised_word2index
+        # self.index2word = revised_index2word
+
+        logger.info("Embedding index build complete")
+
+    def preprocess(self, sentence):
+        """Preprocess an input string to allow only alphabets and numbers
```

Comment:
We shouldn't do pre-processing in model, input is list of tokens

Reply:
`sl_vocab.py` doesn't host any model as such. It is just a utility script to extract the data and stream it in an easy way. It is a "model" for the generator.
The idea was user just gives path to dataset and the model does everything.
Should I still remove preprocessing?